File: python/numerical-stacking-puzzle/midsnd.py
"""
Copyright 2026 Miyabi

Project Name  : Numerical Stacking Puzzle (Python)
File Name     : midsnd.py
Author        : Miyabi
Create        : 2026/06/01
Purpose       : 音再生のミドルウェア
"""

#
# Import
#
from __future__ import annotations

from PySide6.QtMultimedia import QSoundEffect
from PySide6.QtCore import QUrl

#
# Definition
#

# 音ID
E_MIDSND_STACK   = 0    # 音ID:スタック音
E_MIDSND_MERGE_1 = 1    # 音ID:マージ音1
E_MIDSND_MERGE_2 = 2    # 音ID:マージ音2
E_MIDSND_MERGE_3 = 3    # 音ID:マージ音3
E_MIDSND_NUM     = 4    # 音IDの総数

#
# Definition of Class
#
class MidSnd:
    _inst = None

    @staticmethod
    def get_inst():
        return MidSnd._inst

    #----- Constructor(s)/Destructor
    def __init__(self):
        """コンストラクタ
        """
        #------------------------------------------------------------------------
        # インスタンスポインタ保存
        MidSnd._inst = self

        #------------------------------------------------------------------------
        # 初期値設定
        self.disable_play = False   # 再生禁止FG <- OFF

        #------------------------------------------------------------------------
        # バッファクリア
        self._snd_set = [False] * E_MIDSND_NUM
        # QSoundEffectインスタンス
        self._snd = [None] * E_MIDSND_NUM

        self._init_sound()

    def _init_sound(self):
        """
        wavファイルをロード
        同じフォルダに置く想定
        """

        filelist = [
            "stack.wav",
            "merge1.wav",
            "merge2.wav",
            "merge3.wav",
        ]

        for i in range(E_MIDSND_NUM):
            snd = QSoundEffect()
            snd.setSource(QUrl.fromLocalFile(filelist[i]))
            snd.setVolume(0.25)
            self._snd[i] = snd

    def get_disable_play(self) -> bool:
        """再生禁止FGの取得
        """
        return self.disable_play

    def set_disable_play(self, flag: bool):
        """再生禁止FGの更新
        """
        self.disable_play = bool(flag)

    def set_snd(self, snd_id: int):
        """再生する音声を登録する
    
        Args:
            snd_id=音ID
        """
        if 0 <= snd_id < E_MIDSND_NUM:
            self._snd_set[snd_id] = True

    def update(self) -> bool:
        """定期更新処理

        Returns:
            False=処理終了（オブジェクト破棄）
        
        Notes:
            呼び出し周期＝1/60sec
        """
        if self.disable_play:
            # フラグだけ消す
            self._snd_set = [False] * E_MIDSND_NUM
            return True

        # 音ID毎に再生処理する
        for i in range(E_MIDSND_NUM):
            # 有効な音IDのみ処理する
            if self._snd_set[i]:
                snd = self._snd[i]
                if snd is not None:
                    # 音を非同期で再生
                    snd.stop()
                    snd.play()

        # フラグは1フレームで消える（C++と同じ）
        self._snd_set = [False] * E_MIDSND_NUM

        return True

File: cpp/numerical-stacking-puzzle/midsnd.cpp
/********************************************************************************
* Copyright 2026 Miyabi															*
*																				*
* Project Name	: Numerical Stacking Puzzle (Windows 11, Visual Studio 2022)	*
* File Name		: midsnd.cpp													*
* Author		: Miyabi														*
* Create		: 2026/06/01 													*
* Purpose		: 音再生のミドルウェア											*
********************************************************************************/

/********************************************************************************
*	Include Files																*
********************************************************************************/
#include "framework.h"	/* 必須 */
#include <MMSystem.h>
#include "define.h"
#include "midsnd.h"

extern const u_char F_MIDSND_BUF_STACK[];
extern const u_char F_MIDSND_BUF_MERGE_1[];
extern const u_char F_MIDSND_BUF_MERGE_2[];
extern const u_char F_MIDSND_BUF_MERGE_3[];

/********************************************************************************
*	Defines																		*
********************************************************************************/
// シーケンス番号
#define	E_SEQ_INIT	(0)	// 初期状態
#define	E_SEQ_IDLE	(1)	// 待機状態

/********************************************************************************
*	Define Static Instance														*
********************************************************************************/
CRMidSnd*	CRMidSnd::m_lpThis;

/********************************************************************************
*	CRMidSnd Public Constructor(s)/Destructor									*
********************************************************************************/

/********************************************************************************
* FunctionName:	CRMidSnd														*
* Function    :	コンストラクタ													*
* Argument(s) :	none															*
* Attention(s): このコンストラクタは使用しない(引数付きで生成する事)			*
********************************************************************************/
CRMidSnd::CRMidSnd(void)
{
	// 引数付きで作成する事
	Assert(FALSE);
}

/********************************************************************************
* FunctionName:	CRMidSnd														*
* Function    :	コンストラクタ													*
* Argument(s) :	hInst=インスタンスハンドル										*
********************************************************************************/
CRMidSnd::CRMidSnd(HINSTANCE hInst)
{
	//-----------------------------------------------------------------------
	// インスタンスポインタ保存
	m_lpThis = this;

	// シーケンス設定（関数ポインタ配列作成）
	m_lpSeq[E_SEQ_INIT] = &CRMidSnd::SeqInit;	// 初期状態
	m_lpSeq[E_SEQ_IDLE] = &CRMidSnd::SeqIdle;	// 待機状態

	//-----------------------------------------------------------------------
	// バッファクリア
	memset(&(m_abSndSet[0]), 0, sizeof(m_abSndSet));

	//-----------------------------------------------------------------------
	// 初期値設定
	m_hAppInst = hInst;		// インスタンスハンドル
	m_bDePlay = FALSE;		// 再生禁止FG <- OFF
	m_nSeq = E_SEQ_INIT;	// 初期シーケンス=初期状態

	// 初期化正常終了
	this->CRoutine::InitOK();

	Debuglog((LPWSTR)(L"CRMidSnd"));
}

/********************************************************************************
* FunctionName:	~CRMidSnd														*
* Function    :	デストラクタ													*
* Argument(s) :	none															*
********************************************************************************/
CRMidSnd::~CRMidSnd(void)
{
	// インスタンスポインタ破棄
	m_lpThis = nullptr;

	Debuglog((LPWSTR)(L"~CRMidSnd."));
}

/********************************************************************************
*	Private Methods																*
********************************************************************************/

/********************************************************************************
* FunctionName:	PlaySnd															*
* Function    :	音の再生処理													*
* Argument(s) :	none															*
* Return      :	none															*
********************************************************************************/
void
CRMidSnd::PlaySnd(void)
{
	static const u_char* const F_SNDNAME[E_MIDSND_NUM] =
	{
		F_MIDSND_BUF_STACK,
		F_MIDSND_BUF_MERGE_1,
		F_MIDSND_BUF_MERGE_2,
		F_MIDSND_BUF_MERGE_3,
	};

	long nSndID;

	// 音ID毎に再生処理する
	for(nSndID = 0; nSndID < E_MIDSND_NUM; nSndID++)
	{
		// 有効な音IDのみ処理する
		if(m_abSndSet[nSndID])
		{
			// 再生許可時のみ実行
			if(m_bDePlay == FALSE)
			{
				// メモリ上の音データを非同期で再生
				(void)::PlaySound((LPCWSTR)F_SNDNAME[nSndID], nullptr, SND_MEMORY | SND_ASYNC);
			}
			m_abSndSet[nSndID] = FALSE;
		}
	}
}

/********************************************************************************
*	Private Methods (Sequence)													*
********************************************************************************/

/********************************************************************************
* FunctionName:	SeqInit															*
* Function    :	シーケンス処理:初期化											*
* Argument(s) :	none															*
* Return      :	FALSE=処理終了（オブジェクト破棄）								*
********************************************************************************/
bool
CRMidSnd::SeqInit(void)
{
	Debuglog((LPWSTR)(L"MidSndInit."));

	// シーケンス -> 待機状態
	m_nSeq = E_SEQ_IDLE;

	return TRUE;
}

/********************************************************************************
* FunctionName:	SeqIdle															*
* Function    :	シーケンス処理:待機状態											*
* Argument(s) :	none															*
* Return      :	FALSE=処理終了（オブジェクト破棄）								*
********************************************************************************/
bool
CRMidSnd::SeqIdle(void)
{
	PlaySnd();

	return TRUE;
}

/********************************************************************************
*	Public Methods																*
********************************************************************************/

/********************************************************************************
* FunctionName:	Update															*
* Function    :	定期更新処理													*
* Argument(s) :	none															*
* Return      :	FALSE=処理終了（オブジェクト破棄）								*
* Attention(s):	呼び出し周期＝1/60sec											*
********************************************************************************/
bool
CRMidSnd::Update()
{
	//-----------------------------------------------------------------------
	// 基底クラスの処理
	if(this->CRoutine::Update() == FALSE)
	{
		return FALSE;
	}

	//-----------------------------------------------------------------------
	// シーケンス処理
	if((this->*m_lpSeq[m_nSeq])() == FALSE)
	{
		return FALSE;
	}

	return TRUE;
}

/********************************************************************************
* FunctionName:	SetSnd															*
* Function    :	再生する音声を登録する											*
* Argument(s) :	nSndID=音ID														*
* Return      :	none															*
********************************************************************************/
void
CRMidSnd::SetSnd(long nSndID)
{
	m_abSndSet[nSndID] = TRUE;
}

